pkg/controller/volume/vacprotection/vac_protection_controller.go

```diff
@@ -0,0 +1,403 @@
+	if quit {
+		return false
+	}
+	defer c.queue.Done(vacKey)
+
+	vacName := vacKey.(string)
+
+	err := c.processVAC(ctx, vacName)
+	if err == nil {
+		c.queue.Forget(vacKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("VAC %v failed with : %w", vacKey, err))
+	c.queue.AddRateLimited(vacKey)
+
+	return true
+}
+
+func (c *Controller) processVAC(ctx context.Context, vacName string) error {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Processing VAC", "VAC", klog.KRef("", vacName))
+	startTime := time.Now()
+	defer func() {
+		logger.V(4).Info("Finished processing VAC", "VAC", klog.KRef("", vacName), "cost", time.Since(startTime))
+	}()
+
+	vac, err := c.vacLister.Get(vacName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(4).Info("VAC not found, ignoring", "VAC", klog.KRef("", vacName))
+			return nil
+		}
+		return err
+	}
+
+	if protectionutil.IsDeletionCandidate(vac, volumeutil.VACProtectionFinalizer) {
+		// PV should be deleted. Check if it's used and remove finalizer if
+		// it's not.
+		isUsed := c.isBeingUsed(ctx, vac)
+		if !isUsed {
+			return c.removeFinalizer(ctx, vac)
+		}
+		logger.V(4).Info("Keeping VAC because it is being used", "PVC", klog.KRef("", vacName))
+	}
+
+	if protectionutil.NeedToAddFinalizer(vac, volumeutil.VACProtectionFinalizer) {
+		return c.addFinalizer(ctx, vac)
+	}
+
+	return nil
+}
+
+func (c *Controller) addFinalizer(ctx context.Context, vac *storagev1alpha1.VolumeAttributesClass) error {
+	vacClone := vac.DeepCopy()
+	vacClone.ObjectMeta.Finalizers = append(vacClone.ObjectMeta.Finalizers, volumeutil.VACProtectionFinalizer)
+	_, err := c.client.StorageV1alpha1().VolumeAttributesClasses().Update(ctx, vacClone, metav1.UpdateOptions{})
+	logger := klog.FromContext(ctx)
+	if err != nil {
+		logger.V(3).Info("Error adding protection finalizer to VAC", "VAC", klog.KObj(vac), "err", err)
+		return err
+	}
+	logger.V(3).Info("Added protection finalizer to VAC", "VAC", klog.KObj(vac))
+	return nil
+}
+
+func (c *Controller) removeFinalizer(ctx context.Context, vac *storagev1alpha1.VolumeAttributesClass) error {
+	vacClone := vac.DeepCopy()
+	vacClone.ObjectMeta.Finalizers = slice.RemoveString(vacClone.ObjectMeta.Finalizers, volumeutil.VACProtectionFinalizer, nil)
+	_, err := c.client.StorageV1alpha1().VolumeAttributesClasses().Update(ctx, vacClone, metav1.UpdateOptions{})
+	logger := klog.FromContext(ctx)
+	if err != nil {
+		logger.V(3).Info("Error removing protection finalizer from VAC", "VAC", klog.KObj(vac), "err", err)
+		return err
+	}
+	logger.V(3).Info("Removed protection finalizer from VAC", "VAC", klog.KObj(vac))
+	return nil
+}
+
+func (c *Controller) isBeingUsed(ctx context.Context, vac *storagev1alpha1.VolumeAttributesClass) bool {
+	logger := klog.FromContext(ctx)
+
+	pvs, err := c.getPVsAssignedToVAC(vac.Name)
+	if err != nil {
+		logger.Error(err, "Error getting PVs assigned to VAC", "VAC", klog.KObj(vac))
+		return true
+	}
+	if len(pvs) > 0 {
```

Comment:
Should we check pvcs first or pvs first? Does the order matter here?

Reply:
The PVC object is deleted before pv object. So, checking the pv object firstly may accelerate to find the answer.